include/wil/registry_helpers.h

```diff
@@ -0,0 +1,1030 @@
+                {
+                case key_access::read:
+                    return KEY_READ;
+                case key_access::readwrite:
+                    return KEY_ALL_ACCESS;
+                }
+                FAIL_FAST();
+            }
+
+#if defined(_VECTOR_) && defined(_STRING_) && defined(WIL_ENABLE_EXCEPTIONS)
+            template <class InputIt>
+            ::std::vector<wchar_t> get_multistring_from_wstrings(const InputIt& first, const InputIt& last)
+            {
+                ::std::vector<wchar_t> multistring;
+
+                if (first == last)
+                {
+                    multistring.push_back(L'\0');
+                    multistring.push_back(L'\0');
+                    return multistring;
+                }
+
+                for (const auto& wstr : ::wil::make_range(first, last))
+                {
+                    if (!wstr.empty())
+                    {
+                        // back_inserter returns an iterator to enable ::std::copy to write
+                        // from begin() to end() at the back_inserter iterator, growing the vector as necessary
+                        ::std::copy(::std::begin(wstr), ::std::end(wstr), ::std::back_inserter(multistring));
+                    }
+                    multistring.push_back(L'\0');
+                }
+                // double-null-terminate the last string
+                multistring.push_back(L'\0');
+                return multistring;
+            }
+
+            inline ::std::vector<::std::wstring> get_wstring_vector_from_multistring(const ::std::vector<wchar_t>& data)
+            {
+                ::std::vector<::std::wstring> strings;
+                if (data.size() < 3)
+                {
+                    return { ::std::wstring(data.data()) };
+                }
+
+                auto current = data.cbegin();
+                const auto end_iterator = data.cend();
+                const auto lastNull = (end_iterator - 1);
+                while (current != end_iterator)
+                {
+                    const auto next = ::std::find(current, end_iterator, L'\0');
+                    if (next != end_iterator)
+                    {
+                        if (next != lastNull)
+                        {
+                            // don't add an empty string for the final 2nd-null-terminator
+                            strings.emplace_back(current, next);
+                        }
+                        current = next + 1;
+                    }
+                    else
+                    {
+                        current = next;
+                    }
+                }
+                return strings;
+            }
+#endif // #if defined(_VECTOR_) && defined(_STRING_) && defined(WIL_ENABLE_EXCEPTIONS)
+
+#if defined(__SDDL_H__)
+            template <typename err_policy>
+            ::wil::unique_hlocal_ptr<SECURITY_DESCRIPTOR> create_security_descriptor(_In_opt_ PCWSTR security_descriptor)

```

Comment:
I'm not a huge fan of introducing aspects like this---I recommend we take a local pointer to a SECURITY_DESCRIPTOR and leave any fancy stuff to other PRs. I don't really think creating security descriptors is in scope for this PR.

Reply:
the developer wanting to use a security descriptor is going to generally be working with SDDL strings. this makes the API in sync with what developers are already doing in their code.
I'm ok just completely removing support for security descriptors if there is strong pushback here.

Reply:
We removed the SDDL variants :)